src/dp_vnf.c

```diff
@@ -89,83 +156,52 @@ int dp_del_vnf(const uint8_t ul_addr6[DP_VNF_IPV6_ADDR_SIZE])
-			|| (vnf->alias_pfx.ol.ip_type == RTE_ETHER_TYPE_IPV6 && rte_rib6_is_equal(vnf->alias_pfx.ol.ipv6, prefix->ipv6)))
-		&& vnf->alias_pfx.length == prefix_len;
+	struct dp_vnf vnf = {
+		.port_id = port_id,
+		.vni = vni,
+		.type = DP_VNF_TYPE_LB_ALIAS_PFX,
+		.alias_pfx.length = prefix_len,
+		.alias_pfx.ol = *prefix_ip,
+	};
+
+	return dp_vnf_value_exists(&vnf);
 }
 
-bool dp_vnf_lbprefix_exists(uint16_t port_id, uint32_t vni, struct dp_ip_address *prefix_ip, uint8_t prefix_len)
+int dp_del_vnf_by_value(struct dp_vnf *target_vnf)
 {
-	struct dp_vnf *vnf;
-	uint32_t iter = 0;
+	uint8_t *vnf_ul_addr6;
 	int32_t ret;
-	const void *key;
 
-	while ((ret = rte_hash_iterate(vnf_handle_tbl, &key, (void **)&vnf, &iter)) != -ENOENT) {
-		if (DP_FAILED(ret)) {
-			if (prefix_ip->ip_type == RTE_ETHER_TYPE_IPV4)
-				DPS_LOG_WARNING("Iterating VNF table failed", DP_LOG_RET(ret), DP_LOG_PORTID(port_id),
-								DP_LOG_VNI(vni), DP_LOG_IPV4(prefix_ip->ipv4), DP_LOG_PREFLEN(prefix_len));
-			else
-				DPS_LOG_WARNING("Iterating VNF table failed", DP_LOG_RET(ret), DP_LOG_PORTID(port_id),
-								DP_LOG_VNI(vni), DP_LOG_IPV6(prefix_ip->ipv6), DP_LOG_PREFLEN(prefix_len));
-			return false;
-		}
-		if (dp_vnf_match(vnf, DP_VNF_TYPE_LB_ALIAS_PFX, port_id, vni, prefix_ip, prefix_len))
-			return true;
+	ret = rte_hash_lookup_data(vnf_value_tbl, target_vnf, (void **)&vnf_ul_addr6);
+	if (DP_FAILED(ret)) {
+		DP_PRINT_VNF_OP_WARNING("VNF value key lookup failed", target_vnf)
+		return DP_ERROR;
 	}
-	return false;
-}
-
-int dp_del_vnf_by_value(enum dp_vnf_type type, uint16_t port_id, uint32_t vni, struct dp_ip_address *prefix_ip, uint8_t prefix_len)
-{
-	struct dp_vnf *vnf;
-	uint32_t iter = 0;
-	int32_t ret;
-	const void *key;
-	int delete_count = 0;
 
-	while ((ret = rte_hash_iterate(vnf_handle_tbl, &key, (void **)&vnf, &iter)) != -ENOENT) {
-		if (DP_FAILED(ret))
-			return DP_GRPC_ERR_ITERATOR;
+	ret = dp_del_vnf(vnf_ul_addr6);
+	if (DP_FAILED(ret))
+		return ret;
 
-		if (dp_vnf_match(vnf, type, port_id, vni, prefix_ip, prefix_len)) {
-			delete_count++;
-			rte_free(vnf);
-			// this seems unsafe (deletion during traversal), but should be convered by having the table big enough
-			// (ensured by dp_create_jhash_table())
-			// should only ever fail on no-entry or invalid-arguments, but both are covered by rte_hash_iterate()
-			ret = rte_hash_del_key(vnf_handle_tbl, key);
-			if (DP_FAILED(ret)) {
-				if (prefix_ip->ip_type == RTE_ETHER_TYPE_IPV4)
-					DPS_LOG_ERR("Cannot delete VNF key", DP_LOG_RET(ret), DP_LOG_VNF_TYPE(type), DP_LOG_PORTID(port_id),
-								DP_LOG_VNI(vni), DP_LOG_IPV4(prefix_ip->ipv4), DP_LOG_PREFLEN(prefix_len));
-				else
-					DPS_LOG_ERR("Cannot delete VNF key", DP_LOG_RET(ret), DP_LOG_VNF_TYPE(type), DP_LOG_PORTID(port_id),
-								DP_LOG_VNI(vni), DP_LOG_IPV6(prefix_ip->ipv6), DP_LOG_PREFLEN(prefix_len));
-			}
-		}
-	}
-	return delete_count > 0 ? DP_GRPC_OK : DP_GRPC_ERR_NOT_FOUND;
+	return DP_OK;
```

Comment:
You can actually just `return dp_del_vnf(vnf_ul_addr6);`

But saying that, this will in the end do another lookup into the `vnf_value_tbl`. So maybe splitting the `dp_del_vnf()` into `static dp_del_vnf_only()` that gets called here is in order (and then deleting from `vnf_value_tbl` explicitely here.

Reply:
it is really a refactoring effort here. It was a tricky to refactor as you suggested (splitting), since this function is heavily used (e.g., in grpc part). I would suggest to address as a separate PR.